actors/builtin/miner/partition_state.go

```diff
@@ -816,9 +878,99 @@ func (p *Partition) RecordSkippedFaults(
 		return NewPowerPairZero(), NewPowerPairZero(), NewPowerPairZero(), false, xerrors.Errorf("failed to remove recoveries: %w", err)
 	}
 
+	// check invariants
+	if err := p.ValidateState(); err != nil {
+		return NewPowerPairZero(), NewPowerPairZero(), NewPowerPairZero(), false, err
+	}
+
 	return powerDelta, newFaultPower, retractedRecoveryPower, len(newFaultSectors) > 0, nil
 }
 
+// Test that invariants about partition power hold
+func (p *Partition) ValidatePowerState() error {
+	if p.LivePower.Raw.LessThan(big.Zero()) || p.LivePower.QA.LessThan(big.Zero()) {
+		return xerrors.Errorf("Partition left with negative live power: %v", p)
+	}
+
+	if p.UnprovenPower.Raw.LessThan(big.Zero()) || p.UnprovenPower.QA.LessThan(big.Zero()) {
+		return xerrors.Errorf("Partition left with negative unproven power: %v", p)
+	}
+
+	if p.FaultyPower.Raw.LessThan(big.Zero()) || p.FaultyPower.QA.LessThan(big.Zero()) {
+		return xerrors.Errorf("Partition left with negative faulty power: %v", p)
+	}
+
+	if p.RecoveringPower.Raw.LessThan(big.Zero()) || p.RecoveringPower.QA.LessThan(big.Zero()) {
+		return xerrors.Errorf("Partition left with negative recovering power: %v", p)
+	}
+
+	if p.UnprovenPower.Raw.GreaterThan(p.LivePower.Raw) {
+		return xerrors.Errorf("Partition left with invalid unproven power: %v", p)
+	}
+
+	if p.FaultyPower.Raw.GreaterThan(p.LivePower.Raw) {
+		return xerrors.Errorf("Partition left with invalid faulty power: %v", p)
+	}
+
+	if p.RecoveringPower.Raw.GreaterThan(p.LivePower.Raw) || p.RecoveringPower.Raw.GreaterThan(p.FaultyPower.Raw) {
+		return xerrors.Errorf("Partition left with invalid recovering power: %v", p)
+	}
+
+	return nil
+}
+
+// Test that invariants about sector bitfields hold
+func (p *Partition) ValidateBFState() error {
+	// Merge unproven, recoveries, and faults for checks
+	merge, err := bitfield.MultiMerge(p.Unproven, p.Recoveries, p.Faults)
```

Comment:
Recoveries must be contained in faults (checked below), so exclude recoveries here.